### DISCUSSION TITLE
Valid Keycodes Fail to Work for Remote

### DISCUSSION THREAD

**Discussion by liquidarts** [GerryDazoo/Slinger#614] [Q&A]
Briefly put, I'm unable to get a handful of buttons keycodes to work for my device, but I believe I do have the correct codes. Interestingly, the codes that don't work, aren't shown in the terminal output keycodes=[] section.

I've enabled the custom remote config, and confirmed that it's actively being used by changing and observing the codes used in the server terminal to be changing. I've found multiple resources with a list of relevant keycodes from my device from multiple forums online. Having trouble with what would be the Recordings button & Menu, for example. They should be codes 20 & 34, respectively.

I've tried changing the code=1 in the [Remote] section 0 through 3 with no luck.

Happy for any assistance! What I'm not clear is, should I keep testing different values until I stumble on the correct code? Or is the 'keycode =[...]' output relevant? If so, what does it mean if the correct codes are missing?

Also, Gerry, my family greatly appreciates your work on this!

Terminal output
```
Slinginfo  "350/500/M1/M2" 5 30 118.241.117.112 5201 3072 10
8086 Starting Stream for  127.0.0.1:60755
Connecting... ('118.241.117.112', 5201) Control
8086 Selecting Video Source 1
Keycodes= [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 34, 36, 37, 38, 39, 40, 41, 42, 44, 47, 48, 49, 50, 51, 52, 53, 54, 55, 62, 63, 64, 66, 67, 68, 70, 71]
VideoParameters: Resolution= 5 FrameRate= 30 VideoBandwidth= 1000 VideoSmoothness= 63 IframeRate= 5 AudioBitRate= 64
```

Config.ini - Relevant sections

```
...
;500  0=composite, 1=component, 2=HDMI
; If you don't set this the code will use the last configired input
; I recommend setting this value. If not set correctly you'll often only see a black screen. 
VideoSource = 1

[SERVER]
; local port number for the server to listen on for connections
port=8086
;maxstreams=10
enableremote=yes

[REMOTE]
; see release notes for changing this if your remote doesn't work, But as a start make it 
;the same as your configured VideoSource 0-3. See above.
code=1
; path to an external remote control definitions file, if the defaults
; aren't working for you or you want to change the format of the web page.
; use the supplied remote.txt as a starting point.
include=remote.txt 
```

General info
```
SlingBox M2
DVR is Panasonic DMR-BRW520, Japanese model
```

Remote Control codes
```
{"F":"F_DIGIT_0,F_DIG_10","c":"Numeric Keypad","n":"10","K":"018"},
{"F":"F_DIG_12","c":"Numeric Keypad","n":"12","K":"019"},
{"F":"F_RECORDED_TV","c":"Navigation","n":"録画一覧","O":"D-PAD","i":"UNKNOWN","K":"020"},     <-- This is Recording, 20 isn't present in keycodes & fails to work.
{"F":"F_DIG_11","c":"Numeric Keypad","n":"11","K":"021"},
{"F":"F_DIGITAL","c":"Misc","n":"デジタル","i":"UNKNOWN","K":"022"},
{"F":"F_BS","c":"Misc","n":"BS","i":"UNKNOWN","K":"023"},
{"F":"F_PLAY","s":"P","c":"Playback","n":"再生","i":"PLAY","K":"024"},
{"F":"F_STOP","s":"S","c":"Playback","n":"停止","i":"STOP","K":"025"},
{"F":"F_PAUSE","s":"U","c":"Playback","n":"一時停止","i":"PAUSE","K":"026"},
{"F":"F_REWIND","s":"R","c":"Playback","n":"早戻し","i":"REWIND","K":"027"},
{"F":"F_FFWD","s":"F","c":"Playback","n":"早送り","i":"FFWD","K":"028"},
{"F":"F_RECORD","s":"C","c":"Playback","n":"録画","i":"RECORD","K":"029"},
{"F":"F_NEXT_CHAPTER","s":")","c":"Playback","n":"スキップ+","i":"NEXT_CHAPTER","K":"030"},
{"F":"F_PREV_CHAPTER","s":"(","c":"Playback","n":"スキップ-","i":"PREV_CHAPTER","K":"031"},
{"F":"F_BLUE","c":"Teletext","n":"青","i":"BLUE","K":"032"},
{"F":"F_SETUP","s":"N","c":"Menu","n":"Setup","i":"SETUP","K":"033"},     <-- 33 isn't present in keycodes & fails to work.
{"F":"F_CS","c":"Misc","n":"CS","i":"UNKNOWN","K":"034"},

```

**Comment by GerryDazoo** [comment]
The server code lists valid key codes that have been programmed into your  slingbox. So if it's not in the list, it's not valid has you have determined. All I can say is the remote definition file is not what your box has been configured with.
The code= info selects the remote for a particular video source if you have multiple different remotes configured and you want to use the HDMI remote definition when watching the component input for example. In general it should not be needed. It's a hold over from when I didn't have a clear understanding of the remote API when I was reverse engineering the Slingbox protocol.

**Reply by liquidarts** [reply]
Thanks for the info, I appreciate the quick response!
Regarding the 'Keycodes' output when you launch the server, is that coming from the Slingbox?
If so, and if it appears to be misconfigured, is there a way I can change that? I appreciate that this isn't really related to Slinger per se, but I feel like I'm missing something obvious in my setup. For what it's worth, when we used the official apps, the remote control functions worked as expected.
